tools/points-service/main.go

```diff
@@ -0,0 +1,529 @@
+	return db, nil
+}
+
+// --------------------------------------
+//   2) POINTS COMPUTATION LOGIC
+// --------------------------------------
+
+// computePointsForMonths returns a final total if the validator has completed
+// M full months (0-based). If partial month => no increment for that month.
+func computePointsForMonths(blocksActive int64) int64 {
+	// number of fully completed months
+	fullMonths := blocksActive / blocksInOneMonth
+	if fullMonths < 1 {
+		// no full month completed => 0
+		return 0
+	}
+	if fullMonths > int64(len(monthlyIncrements)) {
+		// clamp if user is beyond the last entry
+		fullMonths = int64(len(monthlyIncrements))
+	}
+	// monthlyIncrements is cumulative, so for fullMonths=2 => monthlyIncrements[1] (i.e. 1800)
+	return monthlyIncrements[fullMonths-1]
+}
+
+// updatePoints locks the DB in a transaction and recomputes points for each active row.
+func updatePoints(db *sql.DB, logger *slog.Logger, currentBlock uint64) error {
+	rwLock.Lock()
+	defer rwLock.Unlock()
+
+	tx, err := db.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer func() {
+		if p := recover(); p != nil {
+			_ = tx.Rollback()
+			panic(p)
+		} else if err != nil {
+			_ = tx.Rollback()
+		} else {
+			commitErr := tx.Commit()
+			if commitErr != nil {
+				logger.Error("failed to commit transaction", "error", commitErr)
+			}
+		}
+	}()
+
+	rows, queryErr := tx.Query(`
+        SELECT id, opted_in_block
+        FROM events
+        WHERE opted_out_block IS NULL
+    `)
+	if queryErr != nil {
+		err = fmt.Errorf("failed to query events for points: %w", queryErr)
+		return err
+	}
+	defer rows.Close()
+	// Count rows first
+	var count int
+	countErr := tx.QueryRow("SELECT COUNT(*) FROM events WHERE opted_out_block IS NULL").Scan(&count)
+	if countErr != nil {
+		logger.Error("failed to count active validators", "error", countErr)
+	}
+	logger.Info("updating points",
+		"current_block", currentBlock,
+		"active_validators", count)
+
+	for rows.Next() {
+		var id int64
+		var inBlock uint64
+		if scanErr := rows.Scan(&id, &inBlock); scanErr != nil {
+			logger.Error("scan error", "error", scanErr)
+			continue
+		}
+
+		var blocksActive int64
+		if currentBlock > inBlock {
+			blocksActive = int64(currentBlock - inBlock)
+		}
+		totalPoints := computePointsForMonths(blocksActive)
+
+		_, updErr := tx.Exec(`
+            UPDATE events
+            SET points_accumulated = ?
+            WHERE id = ?
+        `, totalPoints, id)
+		if updErr != nil {
+			logger.Error("failed to update points", "error", updErr, "id", id)
+		}
+	}
+	if rowsErr := rows.Err(); rowsErr != nil {
+		err = fmt.Errorf("rows iteration error: %w", rowsErr)
+	}
+
+	return err
+}
+
+// StartPointsRoutine spawns a background goroutine that runs every 'interval'
+// and calls updatePoints with the latest block.
+func StartPointsRoutine(db *sql.DB, logger *slog.Logger, interval time.Duration, ethClient *ethclient.Client) {
```

Comment:
This routine can be improved. First we should pass in a context to handle clean shutdowns. Second the initial run and ticker runs can be combined. Also in case of errors we should retry with smaller delay.

```suggestion
func StartPointsRoutine(ctx context.Context, db *sql.DB, logger *slog.Logger, interval time.Duration, ethClient *ethclient.Client) {

    ticker := time.NewTicker(interval)
    
    for {
        latestBlock, err := ethClient.BlockByNumber(ctx, nil)
	if err != nil {
		logger.Error("cannot fetch latest block", "error", err)
		ticker.Reset(errorDelay)
	} else {
		currBlockNum := latestBlock.NumberU64()
		// Ideally ctx should be passed in this function as well and checked in the update function.
		if err := updatePoints(db, logger, currBlockNum); err != nil {
			logger.Error("initial points accrual run failed", "error", err)
			ticker.Reset(errDelay)
		} else {
			logger.Info("initial points accrual run completed successfully")
			ticker.Reset(interval)
		}
	}
	
	select {
	    case <-ticker.C:
	    case <-ctx.Done():
	        return
	}
    }
```